### ISSUE TITLE
Please address CurTime's lack of precision

### ISSUE THREAD

**Issue by xaviergmail** [Facepunch/garrysmod-issues#2502]
Player movement is heavily affected by CurTime. It's a float, so precision isn't that great.

After just 9 hours of uptime, we already lose half of our precision:
![](http://xavie.ru/i/VYXQW6)

Compared to 194 days if it were a double
![](http://xavie.ru/i/VYWxuf)

It is not uncommon to have a server stay up for longer than 9/18 hours without a map change, depending on the gamemode.

I know I'm not only speaking for myself when I say this, but scheduling server restarts / map changes just hurts the flow of my gamemode as well as its player count (loading a large map / navmesh can take sometimes up to 8 minutes and people give up trying to reconnect)

```
07:51:23 - [RCON Response] <Xavier> : > print(CurTime())...
524243.21875

07:51:38 - [RCON Response] <Xavier> : > print(CurTime() / 60 / 60 / 24)  -- Days...
6.0678052662037
```

**Comment by aStonedPenguin** [comment]
It's extremely annoying and harmful to player counts having to do server restarts every few days due to this. +1

**Comment by Kefta** [comment]
I agree, however, CurTime is used by A LOT of stuff in the engine so adjusting for the extra precision might take time.

**Comment by maximlahaxim** [comment]
Would be a necessary change. However, it would take quite a while to go through the source code and change all instances of math done with curtime to involve doubles.

**Comment by KingofBeast** [comment]
Can I :+1: more than once pls

**Comment by deinemudda32** [comment]
+1

**Comment by TeddiO** [comment]
> It's extremely annoying and harmful to player counts having to do server restarts every few days due to this. +1

While not a fix as such, you can just reload the map to continue business as usual. I have something which does this automatically every day during low-pop hours.

**Comment by xaviergmail** [comment]
Of course, like every request on here, it'll take time to implement / fix.

However, Willox mentioned something about only having to change the implementation of a few things as well as how it is networked. Most compilers can handle mixes of doubles/floats pretty well so I doubt he'd have to change everything. I don't remember exactly what he said, so take my input with a grain of salt.


> While not a fix as such, you can just reload the map to continue business as usual.
Some servers don't drop below 50 players. A map reload / server restart is extremely difficult for continuous gamemodes , as I said at the bottom of my post.

**Comment by TeddiO** [comment]
It's not ideal but unfortunately it's the only way to deal with the situation until a fix is implemented.

**Comment by DrVrej** [comment]
seems ok +1

**Comment by Noiwex** [comment]
i confirm this

**Comment by G-Force007** [comment]
Bumping this up. This seems to still be a huge problem for servers running for multiple hours on end. Causes vehicles/props/players to become jittery too, seems like a physics problem but it always comes back to CurTime()

**Comment by Kefta** [comment]
Change the map every few hours

**Comment by Noiwex** [comment]
Not a solution.

**Comment by Kefta** [comment]
Never said it was

**Comment by xaviergmail** [comment]
!()[https://xavie.ru/i/kU7yme.png]

Any update on this? Seems like it should be a pretty simple refactor these days..

**Comment by Kefta** [comment]
A lot more than a simple refactor since all places its used, including in networking will have to update the precision. For future updating though, perhaps a Time typedef should be added and change all references to that.

**Comment by xaviergmail** [comment]
Has any consideration been given to this?

**Comment by thegrb93** [comment]
Reload the map each day. I actually do it 2x/day to clear out caches and such.

**Comment by WinterPhoenix96** [comment]
RealTime apparently runs into the same issue if you run a server for more than like 30 days, only change-leveling: https://wiki.garrysmod.com/page/Global/RealTime

**Comment by Noiwex** [comment]
How does this happen? Doesn't it reset after changelevel?

**Comment by Noiwex** [comment]
@WinterPhoenix96

**Comment by WinterPhoenix96** [comment]
Surprisingly not. CurTime resets, but RealTime doesn't: https://winteris.moe/share/2019-08-29_22-23-53.mp4

**Comment by mrwallace888** [comment]
Bumping this. Double precision would make this _significantly_ better across the board. It'll be a pain in the ass to go through the code, but it's what matters in the long run that counts.

**Comment by NathanX-S** [comment]
Bump.

**Comment by RaphaelIT7** [comment]
With it now finally starting to be addressed, I'll try to make a list with everything that needs to be updated to make things easier.
I'll mention only the functions if their arguments / return value needs to be adjusted, if something inside of the function itself needs to change, I'll mention it.

@robotboy655 if you change a function/struct/class variable that is missing from here, please feel free to add it so that we can keep track of what was changed and to reduce how much might break for binary modules.

> [!NOTE]
> This list is incomplete and will grow over time / will be updated.

List with all the things that would definitely need changing for this:
- [x] `TIME_TO_TICKS` & `TICKS_TO_TIME` engine/host.h (Isn't confirmed though, pls confirm Rubat)
- [x] `TIME_TO_TICKS` game/shared/shareddefs.h
- [x] `CBaseServer::GetTime`, `GetTickInterval`, `m_flTickInterval`
- - [x] `IServer::GetTime`, `GetTickInterval`
- [x] `CClientState::GetTime`, `m_tickRemainder`
- [x] `C_BasePlayer::GetTimeBase`
- [x] `CPrediction::RunSimulation`
- [x] `CBasePlayer::[Set/Get]TimeBase`
- [x] `CGlobalVarsBase::curtime`, `interval_per_tick`
- [x] `C_BasePlayer::GetFinalPredictedTime`
- [x] `CBaseEntity::[Set/Get]SimulationTime`
- [x] `CBaseEntity::[Set/Get]AnimTime`
- [x] `CBaseEntity::[Set/Get]NextThink` & `CBaseEntity::GetLastThink`
- [x] `C_BaseEntity::Interpolate`
- [x] `CBaseCombatCharacter::[Set/Get]NextAttack`
- - [x] same for `C_BaseCombatCharacter`
- [ ] `IEffects::Time` & `CEffectsClient::Time` & `CEffectsServer::Time`
- [x] `CClientThinkList::SetNextClientThink`
- [x] `CEngineClient::GetLastTimeStamp`, `Time`
- [x] `CClientState::m_flLastServerTickTime`
- [x] `SV_ComputeClientPacks` inside localnetworkbackdoor `flSaveLastServerTickTime` needs to be changed
- [ ] `CEngine::GetCurTime`
- [ ] `CEngineTool::Time` (includes the interface - IEngineTool::Time)
- [ ] `CVEngineServer::Time` (includes the interface - IVEngineServer::Time)

For variables, you can search for `FIELD_TIME` and in most cases you can easily update the variables to a double as they almost always use CurTime.

Other functions/variables that would need updating since they use CurTime:
- [ ] `CBaseEntity::SetPhysicsAttacker`
- - [ ] `CGib::m_flLastPhysicsInfluenceTime`
- [x] `CBaseEntity::SetupBones`
- [x] `CBaseEntity::[Set/Get]GroundChangeTime`
- [ ] `C_Gib::m_flTouchDelta` (variable)
- [ ] `C_PhysPropClientside::m_flTouchDelta`, `m_fDeathTime` (variable)
- [x] `CRagdoll::GetLastVPhysicsUpdateTime`
- [ ] `CBaseAnimating::CalculateIKLocks`, `Dissolve`
- [x] `CBoneSetup::CalcAutoplaySequences`
- [x] `IBoneSetup::AccumulatePose`, `CalcAutoplaySequences`
- [x] `CBoneCache::UpdateBones`, `IsValid`
- [x] `CBoneSetup::AddSequenceLayers`, `AddLocalLayers`
- [x] `CIKContext::Init` (Update the `CIKContext::m_flTime` variable to double too)
- [x] `bonecacheparams_t::curtime` (variable)
- [x] `CBoneCache::m_timeValid` (variable)
- [x] `dlight_t::die` (variable)
- [x] `Beam_t::die` (variable)
- [ ] `DrawBeamFollow` (`die` arguement - this function in beamsegdraw.h)
- [x] `CBaseAnimatedTextureProxy::GetAnimationStartTime` (all functions that inherit this class need changing too)
- [x] `CAI_ShotRegulator::NextShotTime` (also update m_flNextShotTime if the class)
- - [x] `CAI_ShotRegulator::FireNoEarlierThan`
- [ ] `AI_EnemyInfo_t` (a struct that has lots of `time` variables that need to be updated)
- [ ] `CAI_BaseNPC` (has lots of `m_fl[VariableName]Time` vars that need updating & a lot of `Get/Set[something]Time`)
- - [ ] `CAI_BaseNPC::GetLastNavFailTime`
- [ ] `CAI_PlayerAlly::GetTimePlayerStaring`
- - [ ] `CAI_PlayerAlly::m_TimePendingSet`, `m_flNextRegenTime`, `m_flTimePlayerStartStare`, `m_flNextIdleSpeechTime`, `m_flTimeLastRegen` (variable)
- [ ] `CAI_Path::GetStartTime`, `CAI_Path::m_routeStartTime` (variable)
- [x] `CAI_Senses::GetTimeLastUpdate`
- - [x] `CAI_Senses::m_TimeLastLook`, `m_TimeLastLookHighPriority`, `m_TimeLastLookNPCs`, `m_TimeLastLookMisc` (variable)
- [ ] `CAI_SentenceBase::m_flQueueTimeout` (variable)
- [ ] `ConceptHistory_t::timeSpoken` (variable)
- [ ] `CAI_Expresser::GetSemaphoreAvailableTime`, `GetTimeSpeechComplete`, `GetTimeSpokeConcept`
- - [ ] `CAI_Expresser::m_flStopTalkTime`, `m_flStopTalkTimeWithoutDelay`, `m_flBlockedTalkTime`, `m_flLastTimeAcceptedSpeak` (variable)
- [ ] `CEnvWindShared::WindThink`
- [ ] `CEnvWindShared::m_flVariationTime`, `m_flSimTime`, `m_flSwitchTime` (variable)
- [ ] `GetWindspeedAtTime`
- [ ] `CFuncTrain::m_flNextBlockTime` (variable)
- [ ] `C_ClientRagdoll::m_flFrictionModTime`, `m_flFrictionTime` (variable)
- [x] `S_ComputeDelayForSoundtime`
- [x] `AddAngle::starttime` inside client.h (variable)
- [ ] `CLuaNetworkedVars` - iirc variables store when they were last updated using curtime
- [x] `CSimpleSimTimer::m_next`, `CThinkOnceSemaphore::m_lastTime` inside of game/shared/simtimer.h (variable)
- [x] `ProcessQueryCacheUpdate` Variable inside the code `flCurTime` needs to be changed
- [x] `QueryCacheEntry_t::m_flLastUpdateTime` inside game/shared/querycache.h (variable)
- [x] `CParticleSystemMgr::SetLastSimulationTime` including the `m_flLastSimulationTime` that needs to be changed too
- [ ] `penetrateevent_t::startTime`, `timeStamp` (variable)
- [ ] `friction_t::flLastUpdateTime`, `flLastEffectTime` (variable)
- [ ] `fluidevent_t::impactTime` (variable)
- [x] `PhysGetSyncCreateTime` inside the variables need to be updated
- - [ ] `IPhysicsEnvironment::GetNextFrameTime`, `GetSimulationTime` depends on Valve to be updated to double
- - [ ] `PhysGetNextSimTime`

Specific `CNetworkVar` changes:
- [x] `CBaseEntity::m_flAnimTime`, `m_flSimulationTime`, `m_flPrevAnimTime`, `m_flGroundChangeTime` <- these two are not CNetworkVar
- [ ] `CEntityDissolve::m_flStartTime`
- [ ] `CFlare::m_flTimeBurnOut`
- [ ] `ParticleSmokeGrenade::m_flSpawnTime`, `m_FadeStartTime`, `m_FadeEndTime`
- [ ] `CPlayerLocalData::m_flDucktime`, `m_flDuckJumpTime`, `m_flJumpTime`
- [x] `CBaseCombatWeapon::m_flNextPrimaryAttack`, `m_flNextSecondaryAttack`, `m_flTimeWeaponIdle`
- [ ] `CEnvWindShared::m_flStartTime`
- [ ] `fogparams_t::lerptime`
- [ ] `CSprite::m_flScaleTime`

Updating `FIELD_TIME` -> `DECLARE_FIELD_SIZE( FIELD_TIME, sizeof( double ) )`

Net compat breaking changes (all other changes don't seem to be net compat breaking? nice) :
- [ ] Change `SendPropTime` to use the commented out `return SendPropInt( pVarName, offset, sizeofVar, TIME_BITS, 0, SendProxy_Time );`
~~It seems to be fully fine & when testing with a HL2 Build I didn't cause many delta's~~ :eyes:
I'll test it later more, as I'm like 99% sure I missed something & it may actually cause delta's since the tick base always changes.
If delta really is such a big issue, I'd suggest using TIME_TO_TICKS & then using 32 bits to network it.

<img width="1054" height="284" alt="Image" src="https://github.com/user-attachments/assets/a4b5b412-2418-45dd-8351-8ed5fcd3cc5c" />

**Comment by bloodycop6385** [comment]
Dream

**Comment by NathanX-S** [comment]
> Dream

Very valuable contribution, thank you sir!
